## Оптимизация выполнения кода, векторизация, Numba

Материалы:
* Макрушин С.В. Лекция 3: Оптимизация выполнения кода, векторизация, Numba
* IPython Cookbook, Second Edition (2018), глава 4
* https://numba.pydata.org/numba-doc/latest/user/5minguide.html

## Задачи для совместного разбора

In [47]:
from numba import njit
import numpy as np
import pandas as pd

In [18]:
path1 = "D:/FinUniver/Технологии обработки больших данных/Семинары/02_pandas/data/"

1. Сгенерируйте массив `A` из `N=1млн` случайных целых чисел на отрезке от 0 до 1000. Пусть `B[i] = A[i] + 100`. Посчитайте среднее значение массива `B`.

In [2]:
A = np.random.randint(0, 1000, 10**6)
A

array([594, 920, 705, ..., 700, 715, 988])

In [5]:
def f1(A):
    acc, cnt = 0, 0
    for ai in A:
        bi = ai + 100
        acc += bi
        cnt += 1
    return acc/ cnt

In [4]:
%timeit f1(A)

324 ms ± 11.6 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)


In [6]:
def f2(A):
    acc, cnt = 0, 0
    for ai in A:
        acc +=ai
        cnt += 1
    acc += cnt*100
    return acc/ cnt

In [7]:
%load_ext line_profiler

In [8]:
%timeit f2(A)

230 ms ± 86.4 ms per loop (mean ± std. dev. of 7 runs, 10 loops each)


2. Создайте таблицу 2млн строк и с 4 столбцами, заполненными случайными числами. Добавьте столбец `key`, которые содержит элементы из множества английских букв. Выберите из таблицы подмножество строк, для которых в столбце `key` указаны первые 5 английских букв.

## Лабораторная работа 3

In [12]:
# !pip install line_profiler

1. В файлах `recipes_sample.csv` и `reviews_sample.csv` (__ЛР 2__) находится информация об рецептах блюд и отзывах на эти рецепты соответственно. Загрузите данные из файлов в виде `pd.DataFrame` с названиями `recipes` и `reviews`. Обратите внимание на корректное считывание столбца(ов) с индексами. Приведите столбцы к нужным типам.

Реализуйте несколько вариантов функции подсчета среднего значения столбца `rating` из таблицы `reviews` для отзывов, оставленных в 2010 году.

A. С использованием метода `DataFrame.iterrows` и без использования срезов таблицы

Б. С использованием метода `DataFrame.iterrows` и с использованием срезов таблицы

В. С использованием метода `DataFrame.mean`

Проверьте, что результаты работы всех написанных функций корректны и совпадают. Измерьте выполнения всех вариантов.

In [19]:
reviews = pd.read_csv(path1 + "reviews_sample.csv", sep=',')
reviews

Unnamed: 0     user_id  recipe_id        date  rating  \
0           370476       21752      57993  2003-05-01       5   
1           624300      431813     142201  2007-09-16       5   
2           187037      400708     252013  2008-01-10       4   
3           706134  2001852463     404716  2017-12-11       5   
4           312179       95810     129396  2008-03-14       5   
5           910362       35106      31322  2003-01-03       4   
6           212649      404333     199579  2006-12-10       5   
7           815389      162888      16067  2005-12-09       5   
8           642377       89831      33715  2007-07-03       5   
9          1023302      308434      11252  2008-12-14       5   
10          264646      151754     285773  2008-09-04       0   
11          488041      439759     154996  2007-11-18       5   
12          868771       37275       7974  2002-04-05       5   
13          126930      379985     150499  2008-08-21       5   
14          765972      110278     114415  2005-10-17       5   
15          447429  2002009481     109536  2018-04-19       0   
16          688527       26885      21171  2005-07-11       5   
17          747740      830551      49200  2009-11-17       5   
18          712856      920074      51209  2011-04-19       5   
19          816767      285039     320714  2012-07-01       5   
20          177896      969481      38249  2009-01-24       4   
21         1013284      168896     213395  2007-03-02       5   
22          763435      157425     171615  2006-06-25       5   
23          257148      180898     122289  2008-02-25       4   
24          930153      115006     148347  2006-12-23       5   
25          115270      383346     328668  2010-12-21       5   
26          356766      114681      42603  2007-02-28       5   
27          960040      831240     300681  2008-05-07       5   
28          800781       58931     299881  2010-06-13       5   
29         1004505       58836      61610  2006-02-01       5   
...            ...         ...        ...         ...     ...   
126666      921862      160977      61389  2012-11-11       5   
126667     1069327      152938      28248  2010-08-04       0   
126668      627862       68357      55285  2003-06-08       5   
126669      468313      314579     138173  2014-06-28       0   
126670      416692      191809     112912  2005-07-07       5   
126671        1332      424680      99672  2008-02-21       5   
126672      214204      186855     220408  2007-04-09       5   
126673      494230      894666      29251  2017-05-09       5   
126674      148686       17803      69195  2003-08-29       5   
126675     1046729      160974     187418  2007-10-15       5   
126676      544504      222564      30317  2014-02-11       5   
126677     1020788       47151      81856  2006-10-15       5   
126678      712825      518572      51209  2009-02-21       5   
126679      243637  2002353868     425348  2018-12-02       5   
126680     1069113       57222     157194  2006-07-22       5   
126681     1016683  1802619619     349246  2014-03-26       5   
126682      532061      335737      36548  2008-04-22       5   
126683     1073005     1626288      58976  2010-06-02       0   
126684      698638     1186693     394518  2009-12-09       0   
126685      663966       95743     296868  2010-01-02       4   
126686      605307       17803      19209  2006-01-31       5   
126687      100999      231110     127827  2005-07-25       4   
126688      262084     2359391     292638  2012-10-25       4   
126689      690102      204452      70522  2005-03-21       5   
126690      344676      724631     314698  2008-10-14       5   
126691     1013457     1270706     335534  2009-05-17       4   
126692      158736     2282344       8701  2012-06-03       0   
126693     1059834      689540     222001  2008-04-08       5   
126694      453285  2000242659     354979  2015-06-02       5   
126695      691207      463435     415599  

In [20]:
recipes = pd.read_csv(path1 + "recipes_sample.csv", sep=',')
recipes

name      id  minutes  \
0                  george s at the cove  black bean soup   44123       90   
1                     healthy for them  yogurt popsicles   67664       10   
2                           i can t believe it s spinach   38798       30   
3                                   italian  gut busters   35173       45   
4               love is in the air  beef fondue   sauces   84797       25   
5                               mennonite  corn fritters   44045       15   
6                                   open sesame  noodles  107229       28   
7                             say what   banana sandwich   95926        5   
8                     1 in canada chocolate chip cookies  453467       45   
9                                 412 broccoli casserole  306168       40   
10                                      25 000 casserole   50662      125   
11                         250 00 chocolate chip cookies  118843       20   
12                                     ara s potato  oup   69190      135   
13               blepandekager   danish   apple pancakes  503475       50   
14                              add in anything  muffins  149593       15   
15                          better than tofu  cheesecake  200148       90   
16                     burek  or feta cheese  phyllo pie  310570       65   
17                   da best  chicago style italian beef   95534      645   
18                             german  barbecued carrots  109818       55   
19                                   healthy  fried rice   66932       25   
20                    i can t believe i made cream puffs  226001       55   
21                      kelly s creamy cheddar pea salad  125195       20   
22                         kicked up  oven fried chicken  141939      345   
23            margarita  dip for fresh  or grilled fruit  250883       70   
24                 middle eastern   twice baked potatoes  120297      110   
25                        pass me another   hot clam dip  147477       15   
26                         pasta  with shrimp   eggplant  223349       50   
27                               pizza pan  potato skins   60938       60   
28                                   quick biscuit bread  302399       20   
29                       secret ingredient  bbq meatloaf  342620       75   
...                                                  ...     ...      ...   
29970               zucchini stuffed with feta and basil   72857       30   
29971         zucchini stuffed with goat cheese and mint  130447       30   
29972        zucchini tart with gruyere cheese and herbs  378799       35   
29973                          zucchini tarte with bacon  335783       80   
29974                              zucchini tomato salsa  321224       20   
29975                               zucchini tomato toss   41095       15   
29976                                      zucchini tots  505053       20   
29977                        zucchini wheat germ cookies  174719       30   
29978                 zucchini with bacon corn   peppers  326105       45   
29979               zucchini with bell pepper and tomato  363362       13   
29980               zucchini with jalapeno monterey jack  320622       10   
29981                                zucchini with pasta   96162       30   
29982                          zucchini with pesto sauce   34481       41   
29983                          zucchini with serrano ham  162411       15   
29984  zucchini  corn  black bean  and jack cheese qu...   36349       30   
29985  zucchini  pineapple  ginger jam kicked up with...  325574       50   
29986                 zucchini  potato and herb fritters  344542       30   
29987    zucchini carrot cake with cream cheese frosting  112591       80   
29988                            zucchini carrot muffins  287490       40   
29989                         zucchini cheddar casserole   74023       50   
29990                       zucchini pineapple l

In [23]:
reviews['date'] = pd.to_datetime(reviews['date'], format="%Y-%m-%d")
reviews.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 126696 entries, 0 to 126695
Data columns (total 6 columns):
Unnamed: 0    126696 non-null int64
user_id       126696 non-null int64
recipe_id     126696 non-null int64
date          126696 non-null datetime64[ns]
rating        126696 non-null int64
review        126679 non-null object
dtypes: datetime64[ns](1), int64(4), object(1)
memory usage: 5.8+ MB


In [24]:
def rating_mean_1():
    s = 0
    k = 0
    for i in reviews.iterrows():
        if i[1].date.year == 2010:
            s += i[1].rating
            k += 1
    return s / k

In [25]:
def rating_mean_2():
    s = 0
    k = 0
    for i in reviews[reviews['date'].dt.year == 2010].iterrows():
        s += i[1].rating
        k += 1
    return s / k

In [26]:
def rating_mean_3():
    return reviews[reviews['date'].dt.year == 2010]['rating'].mean()

In [27]:
%time rating_mean_1()

Wall time: 11 s


4.4544402182900615

In [28]:
%time rating_mean_2()

Wall time: 1.02 s


4.4544402182900615

In [29]:
%time rating_mean_3()

Wall time: 15 ms


4.4544402182900615

2. Какая из созданных функций выполняется медленнее? Что наиболее сильно влияет на скорость выполнения? Для ответа использовать профайлер `line_profiler`. Сохраните результаты работы профайлера в отдельную текстовую ячейку и прокомментируйте результаты его работы.

(*). Сможете ли вы ускорить работу функции 1Б, отказавшись от использования метода `iterrows`, но не используя метод `mean`?

In [34]:
%load_ext line_profiler

The line_profiler extension is already loaded. To reload it, use:
  %reload_ext line_profiler


In [37]:
%lprun -f rating_mean_1 rating_mean_1()

Timer unit: 1e-07 s

Total time: 30.0816 s
File: <ipython-input-24-6b378380bdea>
Function: rating_mean_1 at line 1

Line #      Hits         Time  Per Hit   % Time  Line Contents
==============================================================
     1                                           def rating_mean_1():
     2         1         12.0     12.0      0.0      s = 0
     3         1          7.0      7.0      0.0      k = 0
     4    126697  250182672.0   1974.7     83.2      for i in reviews.iterrows():
     5    126696   46856761.0    369.8     15.6          if i[1].date.year == 2010:
     6     12094    3697024.0    305.7      1.2              s += i[1].rating
     7     12094      79274.0      6.6      0.0              k += 1
     8         1         12.0     12.0      0.0      return s / k

In [38]:
%lprun -f rating_mean_2 rating_mean_2()

Timer unit: 1e-07 s

Total time: 2.99737 s
File: <ipython-input-25-a1287ee197e6>
Function: rating_mean_2 at line 1

Line #      Hits         Time  Per Hit   % Time  Line Contents
==============================================================
     1                                           def rating_mean_2():
     2         1         22.0     22.0      0.0      s = 0
     3         1         13.0     13.0      0.0      k = 0
     4     12095   25377671.0   2098.2     84.7      for i in reviews[reviews['date'].dt.year == 2010].iterrows():
     5     12094    4508647.0    372.8     15.0          s += i[1].rating
     6     12094      87287.0      7.2      0.3          k += 1
     7         1         12.0     12.0      0.0      return s / k

In [39]:
%lprun -f rating_mean_3 rating_mean_3()

Timer unit: 1e-07 s

Total time: 0.0187785 s
File: <ipython-input-26-9c1d08824b22>
Function: rating_mean_3 at line 1

Line #      Hits         Time  Per Hit   % Time  Line Contents
==============================================================
     1                                           def rating_mean_3():
     2         1     187785.0 187785.0    100.0      return reviews[reviews['date'].dt.year == 2010]['rating'].mean()

3. Вам предлагается воспользоваться функцией, которая собирает информацию, сколько отзывов содержат то или иное слово. Измерьте время выполнения этой функции. Сможете ли вы найти узкие места в коде, используя профайлер? Выпишите (словами), что в имеющемся коде реализовано неоптимально. Оптимизируйте функцию и добейтесь значительного прироста в скорости выполнения.

In [41]:
def get_word_reviews_count(df):
    word_reviews = {}
    for _, row in df.dropna(subset=['review']).iterrows():
        recipe_id, review = row['recipe_id'], row['review']
        words = review.split(' ')
        for word in words:
            if word not in word_reviews:
                word_reviews[word] = []
            word_reviews[word].append(recipe_id)
    word_reviews_count = {}
    for _, row in df.dropna(subset=['review']).iterrows():
        review = row['review']
        words = review.split(' ')
        for word in words:
            word_reviews_count[word] = len(word_reviews[word])
    return word_reviews_count

In [42]:
%%time
result_origin = get_word_reviews_count(reviews)

Wall time: 29.5 s


In [43]:
%lprun -f get_word_reviews_count get_word_reviews_count(reviews)

Timer unit: 1e-07 s

Total time: 86.9007 s
File: <ipython-input-41-e691285e0ae1>
Function: get_word_reviews_count at line 1

Line #      Hits         Time  Per Hit   % Time  Line Contents
==============================================================
     1                                           def get_word_reviews_count(df):
     2         1         20.0     20.0      0.0      word_reviews = {}
     3    126680  266776544.0   2105.9     30.7      for _, row in df.dropna(subset=['review']).iterrows():
     4    126679   58350975.0    460.6      6.7          recipe_id, review = row['recipe_id'], row['review']
     5    126679    6705196.0     52.9      0.8          words = review.split(' ')
     6   6918689   29236310.0      4.2      3.4          for word in words:
     7   6792010   42057774.0      6.2      4.8              if word not in word_reviews:
     8    174944    1233633.0      7.1      0.1                  word_reviews[word] = []
     9   6792010   44989360.0      6.6      5.2              word_reviews[word].append(recipe_id)
    10         1          9.0      9.0      0.0      word_reviews_count = {}
    11    126680  278954645.0   2202.0     32.1      for _, row in df.dropna(subset=['review']).iterrows():
    12    126679   34924909.0    275.7      4.0          review = row['review']
    13    126679    6730834.0     53.1      0.8          words = review.split(' ')
    14   6918689   32295692.0      4.7      3.7          for word in words:
    15   6792010   66751473.0      9.8      7.7              word_reviews_count[word] = len(word_reviews[word])
    16         1         11.0     11.0      0.0      return word_reviews_count

In [44]:
def get_word_reviews_count_opt(df):
    word_reviews_count = {}
    for row in df.dropna(subset=['review'])['review'].str.split(' '):
        for word in row:
            if word in word_reviews_count:
                word_reviews_count[word] += 1
            else:
                word_reviews_count[word] = 1
    return word_reviews_count

In [45]:
%%time
result_opt = get_word_reviews_count_opt(reviews)

Wall time: 2.43 s


In [46]:
%lprun -f get_word_reviews_count_opt get_word_reviews_count_opt(reviews)

Timer unit: 1e-07 s

Total time: 11.4552 s
File: <ipython-input-44-240fa494db56>
Function: get_word_reviews_count_opt at line 1

Line #      Hits         Time  Per Hit   % Time  Line Contents
==============================================================
     1                                           def get_word_reviews_count_opt(df):
     2         1         15.0     15.0      0.0      word_reviews_count = {}
     3    126680    9026881.0     71.3      7.9      for row in df.dropna(subset=['review'])['review'].str.split(' '):
     4   6918689   27765669.0      4.0     24.2          for word in row:
     5   6792010   37758303.0      5.6     33.0              if word in word_reviews_count:
     6   6617066   39038694.0      5.9     34.1                  word_reviews_count[word] += 1
     7                                                       else:
     8    174944     961983.0      5.5      0.8                  word_reviews_count[word] = 1
     9         1         36.0     36.0      0.0      return word_reviews_count

3. Напишите несколько версий функции `MAPE` (см. [MAPE](https://en.wikipedia.org/wiki/Mean_absolute_percentage_error)) для расчета среднего процентного отклонения значения рейтинга для отзыва от среднего значения рейтинга для этого отзыва.
    1. Без использования массивов `numpy` и `numba`
    2. Без использования массивов `numpy`, но с использованием `numba`
    3. С использованием массивов `numpy`, но без использования `numba`
    4. C использованием массивов `numpy` и `numba`
    
Измерьте время выполнения каждой из реализаций.

Замечание: удалите из выборки отзывы с нулевым рейтингом.

In [48]:
def MAPE_1(df: pd.DataFrame):
    rating_clear = df[df['rating'] != 0]['rating']
    mean = rating_clear.mean()
    return rating_clear.apply(lambda x: abs(mean - x)).sum() / mean * 100 / rating_clear.count()

In [49]:
%time MAPE_1(reviews)

Wall time: 161 ms


11.079150232267242

In [50]:
def MAPE_3(df: pd.DataFrame):
    rating_clear = df[df['rating'] != 0]['rating'].to_numpy()
    mean = rating_clear.mean()
    return np.absolute(rating_clear - mean).sum() / mean * 100 / rating_clear.shape[0]

In [51]:
%timeit MAPE_3(reviews)

9.19 ms ± 205 µs per loop (mean ± std. dev. of 7 runs, 100 loops each)


In [52]:
@njit
def help_M_4(a: np.array):
    mean = a.mean()
    return np.absolute(a - mean).sum() / mean * 100 / a.shape[0]

In [53]:
def MAPE_4(df: pd.DataFrame):
    rating_clear = df[df['rating'] != 0]['rating'].to_numpy()
    return help_M_4(rating_clear)

In [54]:
%timeit MAPE_4(reviews)

8.95 ms ± 490 µs per loop (mean ± std. dev. of 7 runs, 1 loop each)
